### PR TITLE
ISPN-6276 AdvancedAsyncCacheLoader is now threadsafe

### PR DESCRIPTION
**Pull request by slaskawi**
https://issues.jboss.org/browse/ISPN-3938

Highlights:
* `ConcurrentHashSet` is used as a container for all loaded keys

### CODE REVIEW

File: commons/src/main/java/org/infinispan/commons/util/concurrent/ConcurrentHashSet.java [hunk 1]
```diff
@@ -1,4 +1,4 @@
-package org.infinispan.util.concurrent;
+package org.infinispan.commons.util.concurrent;
```
Comment (wburns) [inline commons/src/main/java/org/infinispan/commons/util/concurrent/ConcurrentHashSet.java]: @tristantarrant guessing we should wait until ISPN 9 before moving this?

**Comment by wburns** [comment]
Just to give more information to this issue.  I had discussed this same issue with @danberindei a while back.  Unfortunately making the set a ConcurrentHashSet is not sufficient to prevent all issues.  See [1].  Although it would prevent the exception, however this would then mask the big problem of possibly only seeing a subset of keys.

[1] https://issues.jboss.org/browse/ISPN-3938

File: commons/src/main/java/org/infinispan/commons/util/concurrent/ConcurrentHashSet.java (hunk 1 above)
Comment (slaskawi) [inline commons/src/main/java/org/infinispan/commons/util/concurrent/ConcurrentHashSet.java]: We agreed to postpone it to ISPN9

Comment (wburns) [inline commons/src/main/java/org/infinispan/commons/util/concurrent/ConcurrentHashSet.java]: Sounds good, if you update the PR I can pull it in. Thanks @slaskawi !

File: core/src/test/java/org/infinispan/persistence/PreloadingWithWriteBehindTest.java [hunk 2]
```diff
@@ -31,18 +49,47 @@ protected EmbeddedCacheManager createCacheManager() throws Exception {
...
    }
 
+   @Test(timeOut = 10_000)
+   public void testIfCanLoadKeysConcurrently() throws Exception {
+      final int numberOfEntriesTested = 1_000;
+      final int numberOfConcurrentThreads = 16;
+
+      IntStream
+              .range(0, numberOfEntriesTested)
+              .forEach(i -> cache.put(new KeyWithCollisions(), "value: " + i));
```
Comment (danberindei) [inline core/src/test/java/org/infinispan/persistence/PreloadingWithWriteBehindTest.java]: Is it just me, or does a regular for loop look better?

File: commons/src/main/java/org/infinispan/commons/util/concurrent/ConcurrentHashSet.java (hunk 1 above)
Comment (danberindei) [inline commons/src/main/java/org/infinispan/commons/util/concurrent/ConcurrentHashSet.java]: @slaskawi We could keep the copy in commons, but it should still be in a separate commit/JIRA.

File: core/src/test/java/org/infinispan/persistence/PreloadingWithWriteBehindTest.java (hunk 2 above)
Comment (slaskawi) [inline core/src/test/java/org/infinispan/persistence/PreloadingWithWriteBehindTest.java]: I have no strong opinion on that. Both would look OK in my opinion :)

File: commons/src/main/java/org/infinispan/commons/util/concurrent/ConcurrentHashSet.java (hunk 1 above)
Comment (slaskawi) [inline commons/src/main/java/org/infinispan/commons/util/concurrent/ConcurrentHashSet.java]: I've created a JIRA (mostly as a reminder) for moving it in ISPN9 ([ISPN-6297](https://issues.jboss.org/browse/ISPN-6297))

**Comment by slaskawi** [comment]
Thanks a lot for the review. Comments addressed. Rebased and repushed.

**Comment by slaskawi** [comment]
I changed JIRA to [ISPN-3938](https://issues.jboss.org/browse/ISPN-3938) since [ISPN-6276](https://issues.jboss.org/browse/ISPN-6276) was closed as a duplicate.

**Comment by gustavonalle** [comment]
Merged, thanks @slaskawi !
